File: Week2/Practical/inc_mpi.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! DESCRIPTION:
! This program increments every element of the array by two.
! Compile: 
!	$mpif90 inc_mpi.f90 -o inc_mpif -fopenmp
! Run:
! 	$mpirun -np 2 ./inc_mpif <array length> <nSamples> 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
program inc_serial
  use omp_lib
  use mpi
  implicit none

  integer :: i, j
  integer(kind=16) :: length, nSamples
  integer :: len_per_process, remainder, mylen_per_process
  integer :: procs, rank, ierr
  real(kind=8), allocatable :: a(:), a_per_process(:)
  integer, allocatable :: recvcounts(:), displs(:)
  real(kind=8) :: start_comp, start_comm, last_comp, last_comm
  real(kind=8) :: maxtime_comp, maxtime_comm
  character(len=32) :: arg

  !Get array length and number of times to run the computation
  if (COMMAND_ARGUMENT_COUNT() /=2) then
    write(*,'(a)')"Wrong usage!!!"
    write(*,'(a)')"Correct usage: "
    CALL get_command_argument(0, arg)
    write(*,'(a)')trim(arg)//" <n>  <NSamples>"
    stop
  endif

  CALL get_command_argument(1, arg)
  read(arg,*) length
  CALL get_command_argument(2, arg)
  read(arg,*) NSamples

  call MPI_Init(ierr)
  call MPI_Comm_size(MPI_COMM_WORLD, procs, ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Whole array allocation in master process
  if(rank .eq. 0) then
    allocate(a(length))
  end if
  allocate(recvcounts(procs))
  allocate(displs(procs))

  !Data distribution to processes
  len_per_process = length/procs
  remainder = mod(length, procs)
  if(rank .lt. remainder) then
    mylen_per_process = len_per_process+1
  else
    mylen_per_process = len_per_process
  end if
  call MPI_AllGather(mylen_per_process, 1, MPI_INTEGER, recvcounts, 1, MPI_INTEGER, MPI_COMM_WORLD, ierr)
  displs(1) = 0
  do i=2, procs
    displs(i) = displs(i-1) + recvcounts(i-1)
  end do

  !Sub-Arrays Allocation and Initialisation at each process
  allocate(a_per_process(mylen_per_process))
  do i=1,mylen_per_process
    a_per_process(i)=0.0
  end do

  !Increment elements by 2
  start_comp=omp_get_wtime()
  do i=1, nSamples
    do j=1,mylen_per_process
      a_per_process(j)=a_per_process(j)+2.0
    end do
  end do
  last_comp=omp_get_wtime()-start_comp
  start_comm=omp_get_wtime()
  call MPI_Gatherv(a_per_process, mylen_per_process, MPI_DOUBLE_PRECISION, &
                   a, recvcounts, displs, MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr) 
  last_comm=omp_get_wtime()-start_comm
  
  !Check correctness
  do i=1,length
    if(rank .eq. 0) then
      if(a(i) .ne. 2.0*nSamples) then
        write(*,*)"Not equal at ", i
        exit
      end if
    end if
  end do

  call MPI_Reduce(last_comp, maxtime_comp, 1, MPI_DOUBLE_PRECISION, MPI_MAX, 0, MPI_COMM_WORLD, ierr);
  call MPI_Reduce(last_comm, maxtime_comm, 1, MPI_DOUBLE_PRECISION, MPI_MAX, 0, MPI_COMM_WORLD, ierr);
  if(rank .eq. 0) then
    write(*,*)'Size= ', length, 'Number of processes= ', procs
    write(*,'(a,f18.7, a)')'Computation Time= ', maxtime_comp/nSamples, 'sec'
    write(*,'(a,f18.7, a)')'Communication Time= ', maxtime_comm, 'sec'
  end if

  deallocate(a_per_process)
  deallocate(recvcounts)
  deallocate(displs)
  if(rank .eq. 0) then
    deallocate(a)
  end if
  call MPI_Finalize(ierr)
end program 

File: Week2/Practical/inc_mpi.c
/*****************************************************************************
* DESCRIPTION:
* This program increments every element of the array by two.
* using mpi.
* Compile: 
*	$mpicc inc_mpi.c -o inc_mpic -fopenmp
* Run:
* 	$mpirun -np 2 ./inc_mpic <array length> <nSamples>
******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <omp.h>

int main (int argc, char* argv[])
{
  int i, j, len_per_process, remainder, mylen_per_process;
  long length, nSamples;
  int size, rank;
  int *recvcounts, *displs;
  double *a, *a_per_process;
  double start_comp, start_comm, end_comp, end_comm, maxtime_comp, maxtime_comm;

  /*Get array length and number of times to run the computation*/
  if (argc!=3) {
    printf("Wrong number of arguments!!! \n");
    printf("usage: %s  <array length> <nSamples>\n",argv[0]);
    return -1;
  }
  length=atoi(argv[1]);
  nSamples=atoi(argv[2]);

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  /*Whole array allocation in master process*/
  if(rank==0){
    a=(double*) malloc (length*sizeof(double));
  }
  recvcounts=(int*) malloc (size*sizeof(int));
  displs=(int*) malloc (size*sizeof(int));

  /*Data distribution to processes*/
  len_per_process = length/size; 
  remainder = length%size;
  mylen_per_process = (rank<remainder)?(len_per_process+1):(len_per_process);
  
  MPI_Allgather(&mylen_per_process, 1, MPI_INT, recvcounts, 1, MPI_INT, MPI_COMM_WORLD);
  displs[0] = 0;
  for(i=1; i<size; i++) {
    displs[i] = displs[i-1] + recvcounts[i-1];
  }

  /*Sub-Arrays Allocation and Initialisation at each process*/
  a_per_process = (double*) malloc (mylen_per_process*sizeof(double));
  for (i=0; i<mylen_per_process; i++){
    a_per_process[i]=0.0;
  }

  /*Increment elements by 2*/
  start_comp=omp_get_wtime(); 
  for (i=0; i<nSamples; i++) {
    for (j=0; j<mylen_per_process; j++){
      a_per_process[j]=a_per_process[j]+2.0;  
    }
  }  
  end_comp=omp_get_wtime()-start_comp;
  start_comm=omp_get_wtime(); 
  MPI_Gatherv(a_per_process, mylen_per_process, MPI_DOUBLE, a, recvcounts, displs, MPI_DOUBLE, 0, MPI_COMM_WORLD); 
  end_comm=omp_get_wtime()-start_comm;

  /*Check correctness*/
  for(i=0; i<length; i++){
    if(rank==0){
      if(a[i]!=2.0*nSamples){
        printf("Not equal at %d\n", i);
        break;
      }
    }
  }

  MPI_Reduce(&end_comp, &maxtime_comp, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
  MPI_Reduce(&end_comm, &maxtime_comm, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
  if(rank==0){
    printf ("Size=%ld, Number of processes=%d\n", length, size);
    printf ("Computation Time= %lf sec, Communication Time=  %lf sec\n", maxtime_comp/nSamples, maxtime_comm);
  }

  free(a_per_process);
  free(recvcounts);
  free(displs);
  if(rank==0){
    free(a);
  }
  MPI_Finalize();
  return 0;
}   
